toto/in_toto_record.py

```diff
@@ -0,0 +1,221 @@
+#!/usr/bin/env python
+"""
+<Program Name>
+  in_toto_record.py
+
+<Author>
+  Lukas Puehringer <lukas.puehringer@nyu.edu>
+
+<Started>
+  Nov 28, 2016
+
+<Copyright>
+  See LICENSE for licensing information.
+
+<Purpose>
+  Provides a command line interface to start and stop in-toto link metadata
+  recording.
+
+  start
+    Takes a step name, a functionary's signing key and optional
+    material paths.
+    Creates a temporary link file containing the file hashes of the passed
+    materials and signs it with the functionary's key under
+    .<step name>.link-unfinished
+
+  stop
+    Takes a step name, a functionary's signing key and optional
+    product paths.
+    Expects a .<step name>.link-unfinished in the current directory signed by
+    the functionary's signing key, adds the file hashes of the passed products,
+    updates the signature and renames the file  .<step name>.link-unfinished
+    to <step name>.link
+
+
+  The implementation of the tasks can be found in runlib.
+
+  Example Usage
+  ```
+  in_toto_record.py start --step-name edit-files --materials . --key bob
+  # Edit files manually ...
+  in_toto_record.py stop --step-name edit-files --products . --key bob
+  ```
+
+"""
+import os
+import sys
+import argparse
+import toto.util
+from toto import runlib
+from toto import log
+from toto.models.link import Link
+
+
+def in_toto_record_start(step_name, key_path, material_list):
+  """
+  <Purpose>
+    Starts creating link metadata for a multi-part in-toto step. I.e.
+    loads link signing private key from disk, records passed materials, creates
+    link meta data object from it, signs it with loaded key and stores it to
+    disk as ".<step_name>.link-unfinished".
+
+  <Arguments>
+    step_name:
+            A unique name to relate link metadata with a step defined in the
+            layout.
+    key_path:
+            Path to private key to sign link metadata (PEM).
+    material_list:
+            List of file or directory paths that should be recorded as
+            materials.
+
+  <Exceptions>
+    None.
+
+  <Side Effects>
+    Loads private key from disk
+    Creates a file with  name ".<step_name>.link-unfinished"
+    Calls sys.exit(1) if an exception is raised
+
+  <Returns>
+    Link object (unfinished)
+
+  """
+
+  unfinished_fn = "." + str(step_name) + ".link-unfinished"
+
+  try:
+    log.doing("Start recording '{0}'...".format(step_name))
+    key = toto.util.prompt_import_rsa_key_from_file(key_path)
+
+    log.doing("record materials...")
+    materials_dict = runlib.record_artifacts_as_dict(material_list)
+
+    log.doing("create preliminary link metadata...")
+    link = runlib.create_link_metadata(step_name, materials_dict)
+
+    log.doing("sign metadata...")
+    link.sign(key)
+
+    log.doing("store metadata to '{0}'...".format(unfinished_fn))
+    link.dump(unfinished_fn)
+
+    return link
```

Comment:
TBH, I did this to inspect the link object in a unittest. But know that you mention it it does not seem so smart anymore.